File: src/server/routes/files.ts
import { Router } from 'express'
import { FileService } from '../services/fileService'

const router = Router()
const fileService = new FileService(process.env.WATCH_DIRECTORY || '../../docs')

/**
 * GET /api/files
 * Retrieve all markdown files with metadata
 */
router.get('/', async (req, res) => {
  try {
    const { path: pathFilter, recursive = 'true', sortBy = 'name', order = 'asc' } = req.query

    const options = {
      pathFilter: pathFilter as string,
      recursive: recursive === 'true',
      sortBy: sortBy as 'name' | 'modified' | 'size',
      order: order as 'asc' | 'desc',
    }

    const result = await fileService.getFiles(options)

    res.json({
      success: true,
      data: {
        files: result.files,
        directories: result.directories,
        totalCount: result.files.length,
      },
      timestamp: new Date().toISOString(),
    })
  } catch (error) {
    console.error('Error in GET /api/files:', error)
    res.status(500).json({
      success: false,
      data: null,
      error: {
        code: 'FILE_SYSTEM_ERROR',
        message: error instanceof Error ? error.message : 'Failed to retrieve files',
      },
      timestamp: new Date().toISOString(),
    })
  }
})

/**
 * GET /api/files/:path
 * Retrieve specific file content and metadata
 */
router.get('/*', async (req, res) => {
  try {
    // Get the file path from the URL (everything after /api/files/)
    const filePath = (req.params as string[])[0] as string

    if (!filePath) {
      return res.status(400).json({
        success: false,
        data: null,
        error: {
          code: 'INVALID_PATH',
          message: 'File path is required',
        },
        timestamp: new Date().toISOString(),
      })
    }

    // Decode URL-encoded path
    const decodedPath = decodeURIComponent(filePath)

    const fileContent = await fileService.getFileContent(decodedPath)

    res.json({
      success: true,
      data: fileContent,
      timestamp: new Date().toISOString(),
    })
  } catch (error) {
    console.error('Error in GET /api/files/:path:', error)

    let statusCode = 500
    let errorCode = 'FILE_SYSTEM_ERROR'
    let message = 'Failed to retrieve file'

    if (error instanceof Error) {
      if (error.message.includes('not found') || error.message.includes('ENOENT')) {
        statusCode = 404
        errorCode = 'FILE_NOT_FOUND'
        message = 'File does not exist'
      } else if (error.message.includes('Access denied')) {
        statusCode = 403
        errorCode = 'ACCESS_DENIED'
        message = 'Access to file is not allowed'
      } else {
        message = error.message
      }
    }

    res.status(statusCode).json({
      success: false,
      data: null,
      error: {
        code: errorCode,
        message,
      },
      timestamp: new Date().toISOString(),
    })
  }
})

export default router

File: src/server/app/api/endpoints/files.py
"""File management API endpoints."""

from urllib.parse import unquote

from fastapi import APIRouter, HTTPException, Query

from app.models.api_models import ApiResponse, FileContent, FilesResponse
from app.services.file_service import FileService

router = APIRouter()


@router.get("", response_model=ApiResponse[FilesResponse])
async def get_files(
    path: str | None = Query(None, description="Target path"),
    recursive: bool = Query(True, description="Recursive search"),
    sort_by: str = Query(
        "name",
        pattern="^(name|modified|size)$",
        description="Sort criteria",
        alias="sortBy",
    ),
    order: str = Query("asc", pattern="^(asc|desc)$", description="Sort order"),
) -> ApiResponse[FilesResponse]:
    """
    Get file list.

    Args:
        path: Target path
        recursive: Recursive search
        sort_by: Sort criteria
        order: Sort order

    Returns:
        File list response
    """
    try:
        file_service = FileService()
        result = await file_service.get_files(
            path=path, recursive=recursive, sort_by=sort_by, order=order
        )

        return ApiResponse(success=True, data=result)

    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"File list retrieval error: {str(e)}"
        ) from e


@router.get("/{file_path:path}", response_model=ApiResponse[FileContent])
async def get_file_content(file_path: str) -> ApiResponse[FileContent]:
    """
    Get file content.

    Args:
        file_path: File path (URL encoded)

    Returns:
        File content response
    """
    try:
        # URL decode
        decoded_path = unquote(file_path)

        file_service = FileService()
        content = await file_service.get_file_content(decoded_path)

        return ApiResponse(success=True, data=content)

    except FileNotFoundError as e:
        raise HTTPException(
            status_code=404, detail=f"File not found: {file_path}"
        ) from e
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"File content retrieval error: {str(e)}"
        ) from e
